// base-ui-components/dist/components/BaseDetailBox/BaseDetailBox.vue.es.js
import o from "../BaseBox/BaseBox.vue.es.js";
import l from "../../mixins/i18n.es.js";
import "./BaseDetailBox.vue.es2.js";
import s from "../../_virtual/_plugin-vue2_normalizer.es.js";
const n = {
  name: "BaseDetailBox",
  components: {
    BaseBox: o
  },
  mixins: [
    l
  ],
  props: {
    /**
     * an array with objects to pass the data to be displayed, expecting an object with
     * `label` and `value` properties - if different please use `valuePropertyName`
     * and `labelPropertyName` respectively!
     */
    data: {
      type: Array,
      default: () => []
    },
    /**
     * customize the value property displayed
     */
    valuePropertyName: {
      type: String,
      default: "value"
    },
    /**
     * customize the label property displayed
     */
    labelPropertyName: {
      type: String,
      default: "label"
    },
    /**
     * specify a language (ISO 639-1) (used for label if label is language specific object
     * e.g. `{ de: 'xxx', en: 'yyy' }`)
     */
    language: {
      type: String,
      default: "en"
    }
  }
};
var i = function() {
  var e = this, a = e._self._c;
  return a("base-box", { staticClass: "base-detail-box", attrs: { "box-size": { width: "100%" }, "box-hover": !1, "box-ratio": "0" } }, [a("dl", { staticClass: "base-detail-box-dl" }, [e._l(e.data, function(t, r) {
    return [a("dt", { key: "dt" + r }, [e._v(" " + e._s(e.getLangLabel(t[e.labelPropertyName], !0)) + " ")]), a("dd", { key: "dd" + r }, [e._v(" " + e._s(t[e.valuePropertyName]) + " ")])];
  })], 2)]);
}, d = [], p = /* @__PURE__ */ s(
  n,
  i,
  d,
  !1,
  null,
  "c88db47e",
  null,
  null
);
const b = p.exports;
export {
  b as default
};
