// base-ui-components/dist/components/BaseImage/BaseImage.vue.es.js
import "lazysizes";
import "./BaseImage.vue.es2.js";
import a from "../../_virtual/_plugin-vue2_normalizer.es.js";
const s = {
  name: "BaseImage",
  props: {
    /**
     * set `alt` attribute
     */
    alt: {
      type: String,
      default: ""
    },
    /**
     * specify lazy image loading
     */
    lazyload: {
      type: Boolean,
      default: !1
    },
    /**
     * set `src` attribute
     */
    src: {
      type: String,
      default: ""
    },
    /**
     * set `srcSet` attribute
     * structure: `[{ '640w': 'path' }]`
     */
    srcset: {
      type: Array,
      default: () => []
    }
  },
  computed: {
    clearPng() {
      return "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8//t3PQAJbAN2AyakNQAAAABJRU5ErkJggg==";
    },
    source() {
      const t = this.srcset.length - 1;
      return this.srcset[t] ? Object.values(this.srcset[t])[0] : this.src;
    },
    sourceSet() {
      return this.srcset.length ? this.srcset.map((t) => Object.keys(t).map((e) => `${t[e]} ${e}`)).join(", ") : null;
    }
  },
  methods: {
    emitError(t) {
      this.$emit("error", t);
    }
  }
};
var l = function() {
  var e = this, r = e._self._c;
  return r("img", { class: ["base-image", { lazyload: e.lazyload }], attrs: { alt: e.alt, "data-src": e.lazyload ? e.source : null, "data-srcset": e.lazyload ? e.sourceSet : null, src: e.lazyload ? e.clearPng : e.source, srcset: e.lazyload ? null : e.sourceSet }, on: { error: e.emitError } });
}, o = [], c = /* @__PURE__ */ a(
  s,
  l,
  o,
  !1,
  null,
  "586875cf",
  null,
  null
);
const i = c.exports;
export {
  i as default
};
