// base-ui-components/dist/components/BaseLoader/BaseLoader.vue.es.js
import { ref as r, watch as l } from "vue";
import { useAnnouncer as s } from "../../composables/useAnnouncer.es.js";
import "./BaseLoader.vue.es2.js";
import i from "../../_virtual/_plugin-vue2_normalizer.es.js";
const u = {
  name: "BaseLoader",
  props: {
    /**
     * define a color for the loader
     */
    loaderColor: {
      type: String,
      default: ""
    },
    /**
     * add positional properties like for a style object
     */
    position: {
      type: Object,
      default: () => ({})
    },
    /**
     * set this to `true` if you want to keep element (with height and width) but dont
     * want the loader to show
     */
    hide: {
      type: Boolean,
      default: !1
    },
    /**
     * define a text that should be read as soon as loader appears
     */
    textOnLoaderShow: {
      type: String,
      default: "loading"
    }
  },
  setup(o) {
    const e = r(null), t = r(null), { announcement: a } = s(e);
    function n() {
      e.value && !o.hide && (t.value && (clearTimeout(t.value), t.value = null), t.value = setTimeout(() => {
        e.value && !o.hide && (a.value = o.textOnLoaderShow);
      }, 1e3));
    }
    return l(e, () => {
      n();
    }), {
      loader: e,
      announcement: a,
      setLoaderAnnouncement: n
    };
  },
  data() {
    return {
      timeout: null
    };
  },
  watch: {
    /**
     * for some reason in frontend setup watcher did not trigger for prop `hide`
     * so added it here
     */
    hide() {
      this.setLoaderAnnouncement();
    }
  }
};
var d = function() {
  var e = this, t = e._self._c;
  return t("div", { ref: "loader", staticClass: "base-loader", style: { ...e.position, "--loader-color": e.loaderColor } }, [e.hide ? e._e() : t("svg", { staticClass: "base-loader-circular" }, [t("circle", { staticClass: "base-loader-path", attrs: { cx: "50", cy: "50", r: "20", fill: "none", "stroke-width": "4", "stroke-miterlimit": "10" } })])]);
}, c = [], f = /* @__PURE__ */ i(
  u,
  d,
  c,
  !1,
  null,
  "980d8d19",
  null,
  null
);
const h = f.exports;
export {
  h as default
};
