// base-ui-components/dist/components/BaseBreadCrumbs/BaseBreadCrumbs.vue.es.js
import l from "../BaseLink/BaseLink.vue.es.js";
import "./BaseBreadCrumbs.vue.es2.js";
import u from "../../_virtual/_plugin-vue2_normalizer.es.js";
const c = {
  name: "BaseBreadCrumbs",
  components: {
    BaseLink: l,
    BaseIcon: () => import("../BaseIcon/BaseIcon.vue.es.js").then((r) => r.default || r)
  },
  props: {
    /**
     * specify an array with the properties `route`, `label` and (optional) `icon` and `showLabel` that
     *  should be used for navigation - these have to be internal
     *  links, i.e. without a protocol like `http:` or `mailto:` in the url
     * `showLabel` set to false is only considered if the `icon` property is set!
     */
    routes: {
      type: Array,
      default: () => [{ route: "", label: "" }],
      validator: (r) => !r.some(
        (e) => e.route && !!(e.route.match(/^([a-z][a-z0-9+\-.]*:\/\/)/) || e.route.match(/^mailto:/) || e.route.match(/^tel:/)) || e.showLabel === !1 && !e.icon
      )
    },
    /**
     * specify how link element should be rendered - this needs to be a
     * valid vue link component (e.g. RouterLink, NuxtLink) and vue-router
     * is necessary
     */
    renderLinkAs: {
      type: String,
      default: "RouterLink"
    }
  }
};
var _ = function() {
  var e = this, a = e._self._c;
  return a("div", { staticClass: "base-bread-crumbs" }, [e._l(e.routes, function({ route: n, label: s, showLabel: t, icon: o }, i) {
    return [a("BaseLink", { key: n, class: ["base-bread-crumbs__link", { "base-bread-crumbs__link--icon-only": t === !1 }], attrs: { "external-link-target": "_self", "identifier-property-value": n, value: s, "render-link-as": e.renderLinkAs, "additional-attributes": { ariaCurrentValue: "page" } }, scopedSlots: e._u([{ key: "label", fn: function() {
      return [a("span", { staticClass: "base-bread-crumbs__label" }, [o ? a("BaseIcon", { key: s, class: [
        "base-bread-crumbs__icon",
        { "base-bread-crumbs__icon--with-label": t !== !1 }
      ], attrs: { name: o, title: t === !1 ? s : void 0 } }) : e._e(), t !== !1 ? a("span", { staticClass: "base-bread-crumbs__label-text" }, [e._v(" " + e._s(s) + " ")]) : e._e()], 1)];
    }, proxy: !0 }], null, !0) }), i < e.routes.length - 1 ? a("BaseIcon", { key: `${n}-separator`, staticClass: "base-bread-crumbs__separator", attrs: { name: "next" } }) : e._e()];
  })], 2);
}, d = [], m = /* @__PURE__ */ u(
  c,
  _,
  d,
  !1,
  null,
  "295785d0",
  null,
  null
);
const k = m.exports;
export {
  k as default
};
