// base-ui-components/dist/components/BaseImageGrid/BaseImageGrid.vue.es.js
import n from "../BaseImage/BaseImage.vue.es.js";
import "./BaseImageGrid.vue.es2.js";
import i from "../../_virtual/_plugin-vue2_normalizer.es.js";
const o = {
  name: "BaseImageGrid",
  components: {
    BaseImage: n
  },
  props: {
    /**
     * specify alt text (used for each image)
     */
    alt: {
      type: String,
      default: ""
    },
    /**
     * specify lazy image loading
     */
    lazyload: {
      type: Boolean,
      default: !0
    },
    /**
     * specify images to display in a grid
     *
     * array variants to render a single image using `src` or `srcset` for a responsive image:
     * **src**: `['path']`
     * **srcset**: `[{ '640w': 'path' }]`
     */
    images: {
      type: Array,
      default: () => []
    }
  },
  computed: {
    /**
     * Reduce images array to the first 4 images
     * @returns {*[]}
     */
    imagesInt() {
      return this.images.slice(0, 4);
    }
  },
  methods: {
    /**
     * get single src
     *
     * @param {string|object} data
     * @returns {string}
     */
    src(e) {
      return typeof e == "string" ? e : Object.entries(e).slice(-1)[0][1];
    },
    /**
     * get srcset
     *
     * @param {object} data
     * @returns {array}
     */
    srcset(e) {
      return typeof e == "object" ? [e] : [];
    }
  }
};
var m = function() {
  var t = this, r = t._self._c;
  return r("div", { class: [
    "base-image-grid",
    "base-image-grid--" + t.imagesInt.length
  ] }, t._l(t.imagesInt, function(s, a) {
    return r("div", { key: "image" + a, staticClass: "base-image-grid__item" }, [r("BaseImage", { attrs: { lazyload: !0, src: t.src(s), srcset: t.srcset(s) } })], 1);
  }), 0);
}, c = [], l = /* @__PURE__ */ i(
  o,
  m,
  c,
  !1,
  null,
  "cbdb3667",
  null,
  null
);
const d = l.exports;
export {
  d as default
};
